### PR TITLE
New Atlona PRO3 binding (#1171)

### PR DESCRIPTION
**Pull request by tmrobert8**
Implementation of Atlona binding

Signed-off-by: Tim Roberts troberts@bigfoot.com (github: tmrobert8)

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Ready to be reviewed?

**Comment by tmrobert8** [comment]
There is a new firmware about to come out (in beta right now) - but it should have minor changes (if any).  I'd say this is probably good to be reviewed

**Comment by kaikreuzer** [comment]
Ok, so please rebase your branch in order to resolve merge conflicts.
Also, as a quick first feedback, please check the code against the guidelines, e.g.

- remove version constraints from imports in the manifest
- update the license headers
- remove the "version" annotations from the Javadoc headers

Thanks!

**Comment by tmrobert8** [comment]
Kai,

Just to let you know - I'm carrying over the comments you've made on the russound one into this one and will commit this week.  You may want to hold off on any further review until the new commit.

Tim

**Comment by tmrobert8** [comment]
Kai,

I've updated the addon with most of the comments from the other addons (the channel groups really were useful on this one).  Feel free to review and rip it apart!  :)

Tim

File: addons/binding/org.openhab.binding.atlona/README.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
+# Atlona Binding
+
+This binding integrates Atlona AT-UHD-PRO3 HdBaseT matrix switches (http://www.atlona.com) into your openHAB installation.
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/README.md]: make it a link: `[Atlona AT-UHD-PRO3 HdBaseT matrix switches](http://www.atlona.com)`

File: addons/binding/org.openhab.binding.atlona/README.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+| Thing         | Thing Type | Description                                             |
+|---------------|------------|---------------------------------------------------------|
+| pro3-44m      | Thing      | The AT-UHD-PRO3-44M 4x4 HDBaseT matrix.                 |
+| pro3-66m      | Thing      | The AT-UHD-PRO3-66M 6x6 HDBaseT matrix.                 |
+| pro3-88m      | Thing      | The AT-UHD-PRO3-88M 8x8 HDBaseT matrix.                 |
+| pro3-1616m    | Thing      | The AT-UHD-PRO3-1616M 16x16 HDBaseT matrix.             |
+
+## Discovery
+
+The Atlona AT-UHD-PRO3 switch can be discovered by starting a scan in the paperui and then logging into your switch and pressing the "SDDP" button on the "Network" tab.  The "SDDP"
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/README.md]: Paper UI

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/discovery/AtlonaDiscovery.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                                    }
+                                } catch (SocketTimeoutException e) {
+                                    // ignore
+                                }
+                            }
+
+                            multiSocket.close();
+                        } catch (Exception e) {
+                            if (!e.getMessage().contains("No IP addresses bound to interface")) {
+                                logger.debug("Error getting ip addresses: {}: {}", e.getMessage(), e);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/discovery/AtlonaDiscovery.java]: Should be:
```
logger.debug("Error getting ip addresses: {}", e.getMessage(), e);
```

Please check the same for all your other logging statements

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/discovery/AtlonaDiscovery.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+            } else if (model.equalsIgnoreCase("AT-UHD-PRO3-88M")) {
+                typeId = AtlonaBindingConstants.THING_TYPE_PRO3_88M;
+            } else if (model.equalsIgnoreCase("AT-UHD-PRO3-1616M")) {
+                typeId = AtlonaBindingConstants.THING_TYPE_PRO3_1616M;
+            } else {
+                logger.warn("Unknown model #: {}");
+            }
+
+            if (typeId != null) {
+                logger.info("Creating binding for {} ({})", model, from);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/discovery/AtlonaDiscovery.java]: reduce to debug

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package org.openhab.binding.russound.internal.net;
+package org.openhab.binding.atlona.internal.net;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java]: Ouch - you are moving this file away from the Russound binding!

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketSession.java [hunk 6]
```diff
@@ -6,7 +6,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-v10.html
  */
-package org.openhab.binding.russound.internal.net;
+package org.openhab.binding.atlona.internal.net;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketSession.java]: same here

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java (hunk 5 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java]: btw, you should make these adaptions in Russound as well, right?

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketSessionListener.java [hunk 7]
```diff
@@ -6,7 +6,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-v10.html
  */
-package org.openhab.binding.russound.internal.net;
+package org.openhab.binding.atlona.internal.net;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketSessionListener.java]: same here

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/pro3/AtlonaPro3Handler.java [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+
+        if (command instanceof RefreshType) {
+            handleRefresh(channelUID);
+            return;
+        }
+
+        // if (getThing().getStatus() != ThingStatus.ONLINE) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/pro3/AtlonaPro3Handler.java]: remove

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/pro3/AtlonaPro3PortocolHandler.java [hunk 9]
```diff
@@ -0,0 +1,1190 @@
...
+    }
+
+    /**
+     * Resets the matrix back to defaults.
+     */
+    void resetMatrix() {
+        sendCommand(CMD_MATRIX_RESET);
+    }
+
+    // void sendRS232Command(int zoneNbr, String rs232Cmd) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/pro3/AtlonaPro3PortocolHandler.java]: remove?

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/pro3/AtlonaPro3PortocolHandler.java [hunk 10]
```diff
@@ -0,0 +1,1190 @@
...
+     * @param resp the possibly null, possibly empty actual response
+     */
+    private void handleMatrixResetResponse(String resp) {
+        if (RSP_MATRIX_RESET.equals(resp)) {
+            _callback.statusChanged(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "System is rebooting due to matrix reset");
+        }
+    }
+
+    // private void handleRs232Response(Matcher m, String resp) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/pro3/AtlonaPro3PortocolHandler.java]: remove?

File: addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+        }
+        _host = host;
+        _port = port;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see
+     * org.openhab.binding.russound.internal.net.SocketSession#addListener(org.openhab.binding.russound.internal.net.
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java]: remove the russound references - probably best to completely remove the JavaDoc on overridden methods as they do not bring much value.

File: addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+     *
+     * @see org.openhab.binding.russound.internal.net.SocketSession#sendCommand(java.lang.String)
+     */
+    @Override
+    public synchronized void sendCommand(String command) throws IOException {
+        if (command == null) {
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        // if (command.trim().length() == 0) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java]: remove?

File: addons/binding/pom.xml [hunk 13]
```diff
@@ -1,72 +1,73 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kaikreuzer) [inline addons/binding/pom.xml]: please revert these changes and only add a single new line

File: addons/binding/org.openhab.binding.atlona/META-INF/MANIFEST.MF [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+ org.apache.commons.lang,
+ org.eclipse.smarthome.config.core,
+ org.eclipse.smarthome.config.discovery,
+ org.eclipse.smarthome.core.library.types,
+ org.eclipse.smarthome.core.thing,
+ org.eclipse.smarthome.core.thing.binding,
+ org.eclipse.smarthome.core.thing.binding.builder,
+ org.eclipse.smarthome.core.thing.type,
+ org.eclipse.smarthome.core.types,
+ org.jupnp.model.meta,
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/META-INF/MANIFEST.MF]: afaics, this is not needed

File: addons/binding/org.openhab.binding.atlona/ESH-INF/thing/thing-types.xml [hunk 15]
```diff
@@ -0,0 +1,712 @@
...
+            <channel id="power" typeId="power"/>
+            <channel id="panellock" typeId="panellock"/>
+            <channel id="irenable" typeId="irenable"/>
+            <channel id="presetcmd" typeId="presetcmd"/>
+            <channel id="matrixcmd" typeId="matrixcmd"/>
+	    </channels>
+    </channel-group-type>  
+	  
+	  
+    <channel-group-type id="portgroup1">
```
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.atlona/ESH-INF/thing/thing-types.xml]: Without the label/description overrides - this becomes a complete mess of not knowing what port is which port (especially on the 16x16) in the paperui and I'd rather not remove it until that issue is resolved.  I'd prefer to do this because it's just alot of duplicate junk that's not needed because of the label/description issue.  What if I just update it when that issue is closed?

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java (hunk 5 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java]: Bit confused - I copied these files from the russound binding and accidentally committed them without changing the namespace on them.  I then updated them to the altona namespace and recommitted (probably should have merged commits as well).  Unless I'm misunderstanding you - it's correct the way it is.

File: addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java (hunk 11 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java]: Yeah - didn't notice that one - thanks!

**Comment by tmrobert8** [comment]
Kai - I believe I've made all the changes except for the channel groups.  I strongly believe that should wait until the label/description override is available since it is really confusing in the Paper UI without them

File: addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java (hunk 5 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding/atlona/internal/net/SocketChannelSession.java]: screen shot 2017-02-02 at 15 10 01

This shows that the file is moved (and edited) and not copied. So if I'd merge this PR, the Russound binding will not contain that class anymore.

File: addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java [hunk 16]
```diff
@@ -0,0 +1,420 @@
+package org.openhab.binding.sony.internal.net;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java]: this file is from the sony binding

**Comment by kaikreuzer** [comment]
>  I strongly believe that should wait until the label/description override is available since it is really confusing in the Paper UI without them

Ok, I agree, let's leave it, but keep in mind to refactor it again, once we can define different labels.

There are still some issues regarding moved files, so please resolve this and we should be good to merge. Btw, you should never merge the master branch into your feature branch, but rather rebase your feature branch on the latest master - this should avoid such issues.

**Comment by tmrobert8** [comment]
Kai - just updated the thing to use label/description overrides - take a look

**Comment by kaikreuzer** [comment]
Thanks! Just checked, but there is still [one file being moved from russound to sony](https://github.com/openhab/openhab2-addons/pull/1252/files#diff-f87f02386361fd7377c6ed41cc8c8ba4) - this must definitely be removed from this PR!

**Comment by tmrobert8** [comment]
Kai - I think I made a mess of my branch.  Thought I had a handle on git up to now - could you take a look and give me some hints on how I might correct all this?

**Comment by tmrobert8** [comment]
BTW - if you take a look at [https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding](https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.atlona/src/main/java/org/openhab/binding)

That's on the master tree and should not be there - I think they snuck in with the russound merge.  Obviously I need to be more careful with git and copying files around.  Not sure how I managed that one either.

**Comment by tmrobert8** [comment]
OK - I think I fixed it up - that atlona branch in the master tree was what was causing all the issues.  The two last commits should have fixed that and pushed the correct files

**Comment by kaikreuzer** [comment]
Ok, looks better now indeed!
But since time has gone by, I have two additional wishes:
1. Please change the version to from 2.0.0 to 2.1.0 (in pom and manifest)
2. Re-run `mvn license:format' to get the latest (2017) headers

**Comment by tmrobert8** [comment]
Done

**Comment by kaikreuzer** [comment]
Cool - let's rock 😎
